parameters/src/testnet3/powers.rs

```diff
@@ -256,112 +202,289 @@ impl<E: PairingEngine> ToBytes for PowersOfG<E> {
+    }
+
+    fn distance_from_shifted_of(&self, range: &Range<usize>) -> usize {
+        (range.start as isize - self.available_powers().1.start as isize).unsigned_abs()
+    }
+
+    /// Assumes that we have the requisite powers.
+    fn shifted_powers(&self, range: Range<usize>) -> Result<&[E::G1Affine]> {
+        ensure!(
+            self.contains_in_shifted_powers(&range),
+            "Requested range is not contained in the available shifted powers"
+        );
+
+        if range.start < MAX_NUM_POWERS / 2 {
+            assert!(self.shifted_powers_of_beta_g.is_empty());
+            // In this case, we have downloaded all the powers, and so
+            // all the powers reside in self.powers_of_beta_g.
+            Ok(&self.powers_of_beta_g[range])
+        } else {
+            // In this case, the shifted powers still reside in self.shifted_powers_of_beta_g.
+            let lower = self.shifted_powers_of_beta_g.len() - (MAX_NUM_POWERS - range.start);
+            let upper = self.shifted_powers_of_beta_g.len() - (MAX_NUM_POWERS - range.end);
+            Ok(&self.shifted_powers_of_beta_g[lower..upper])
+        }
+    }
+
+    /// Assumes that we have the requisite powers.
+    fn normal_powers(&self, range: Range<usize>) -> Result<&[E::G1Affine]> {
+        ensure!(self.contains_in_normal_powers(&range), "Requested range is not contained in the available powers");
+        Ok(&self.powers_of_beta_g[range])
+    }
+
+    /// Returns the power of beta times G specified by `target`.
+    fn power(&mut self, target: usize) -> Result<E::G1Affine> {
+        self.powers(target..(target + 1)).map(|s| s[0])
+    }
+
+    /// Slices the underlying file to return a vector of affine elements between `lower` and `upper`.
+    fn powers(&mut self, range: Range<usize>) -> Result<&[E::G1Affine]> {
+        if range.is_empty() {
+            return Ok(&self.powers_of_beta_g[0..0]);
+        }
+        ensure!(range.start < range.end, "Lower power must be less than upper power");
+        ensure!(range.end <= MAX_NUM_POWERS, "Upper bound must be less than the maximum number of powers");
+        if !self.contains_powers(&range) {
+            // We must download the powers.
+            self.download_powers_for(&range)?;
+        }
+        if self.contains_in_normal_powers(&range) { self.normal_powers(range) } else { self.shifted_powers(range) }
+    }
+
+    pub fn download_powers_for(&mut self, range: &Range<usize>) -> Result<()> {
+        if self.contains_in_normal_powers(range) || self.contains_in_shifted_powers(range) {
+            return Ok(());
+        }
+        let half_max = MAX_NUM_POWERS / 2;
+        if (range.start <= half_max) && (range.end > half_max) {
+            // If the range contains the midpoint, then we must download all the powers.
+            // (because we round up to the next power of two).
+            self.download_up_to(range.end)?;
+            self.shifted_powers_of_beta_g = Vec::new();
```

Comment:
```suggestion
            self.shifted_powers_of_beta_g.clear();
```

Reply:
Doesn't `clear` keeps the old capacity? W e will never use that allocated memory again, which is why used `new` to free that memory ASAP.

Reply:
Yes, will revert